### ISSUE TITLE
bug: not all rows output to Pulsar materialization

### ISSUE THREAD

**Issue by stinkymatt** [kaskada-ai/kaskada#248]
**Description**
Setting up a Pulsar materialization results in fewer 'final' rows being output to the materialization destination than are in the final query. 

**To Reproduce**
1. Load records into a Kaskada table and write a query.
2. Materialize the query to a Pulsar topic
3. Note that the rows in the query don't match the number of rows that are materialized

**Actual Behavior**
In one example (ping me for notebook) what should be a 150 row query result only produces 27 rows in the materialization. Keys towards the top of the input files seem to be more likely to be in the output than keys at the end of the input files.

**Expected Behavior**
The materialization should have the same results as the query

**Additional context**
Ping me on slack for a notebook that reproduces this.

**Relevant Logs / Links**

**Comment by bjchambers** [comment]
Could you post the queries you're using for both cases?

Two things that could be happening:

1. "All results" will have more rows than "final results". Make sure you're comparing two final result queries.
2. If the "final results" query has a "changed since" it may only produce final results that have changed since previously. The idea being that the first time you run a materialization (to a store) you need all of the final results, but after that you only need final results for keys that have changed (so that you can update the store).

**Comment by stinkymatt** [comment]
The problem appears to be happening at the sink to Astra when converting between float types:

`com.datastax.oss.sink.pulsar.CassandraSinkTask - Error decoding/mapping Pulsar record PulsarSinkRecord{SinkRecord(sourceRecord=PulsarRecord(topicName=Optional[persistent://kaskada123/default/materialized], partition=0, message=Optional[org.apache.pulsar.client.impl.MessageImpl@5a8b4dc2], schema=org.apache.pulsar.client.impl.schema.BytesSchema@16acd8fc, failFunction=org.apache.pulsar.functions.source.PulsarSource$$Lambda$566/0x00000008407bf040@30fa7c3f, ackFunction=org.apache.pulsar.functions.source.PulsarSource$$Lambda$565/0x00000008407bfc40@5aa154f8), value=[B@18e85e8c)}: Cannot convert 58874.53999999996 from BigDecimal to Float`

Not sure if that means we should close this as the problem lies elsewhere, or improve the compatibility here. The query in question is:
`%%fenl --result-behavior final-results --var tx_totals
{
  total: transactions.price | with_key(transactions.purchaser_id) | sum()
}`

But it does seem like the math in the sum() is somehow "off-spec" as the source data is 2-5 decimal places.

**Comment by bjchambers** [comment]
> Cannot convert 58874.53999999996 from BigDecimal to Float

This is interesting. I would have expected this conversion to work. Two things though -- it may be converting to java `float` (which is `f32`) and may be better / require `double` (which is `f64` in rust). So it may not have enough bits of precision to do so accurately... but I would have expected some kind of rounding then (eg., `58874.5`, etc.).

I just checked and I can at least confirm that if we produce an `f64` value we report that as `double `in Avro, so this shouldn't be the problem:

https://github.com/kaskada-ai/kaskada/blob/main/crates/sparrow-arrow/src/avro/schema.rs#L74

I also tried this in a playground:

```
import java.math.BigDecimal;

public class Program
{
    public static void main(String[] args) {
        BigDecimal value = new BigDecimal("58874.53999999996");
        double d = value.doubleValue();
        float f = value.floatValue();
	System.out.println("Value: " + value + " d = " + d + " f = " + f);
    }
}
```

And can confirm that at least in Java, the BigDecimal *can* convert that value to both float and double with different precision (Value: 58874.53999999996 d = 58874.53999999996 f = 58874.54).

Probably worth filing a bug elsewhere and linking it to this.

> But it does seem like the math in the sum() is somehow "off-spec" as the source data is 2-5 decimal places.

I don't think that's "off-spec". The mapping of floating point to decimal places is odd -- it's not about how many decimal places it has in "base 10" but how many it has in "base 2". So `0.25` is 1/4 and easy, but `0.3` is hard.

This is why something like `0.1 + 0.1 + 0.1 = 0.30000000000000004`.

So in general, this doesn't seem to me to be "off spec".

Probably OK to close here?
